### PR TITLE
DNN-22048 LockRecursionException

### PR DESCRIPTION
**Pull request by mean2me**
Improved concurrency: introduced UpgradeableReadLock to prevent LockRecursionException

### CODE REVIEW

**Comment by leonardopires** [comment]
Your Pull Request has been selected for a project-agnostic code review by [Aurea Central Reviewers Team](https://github.com/orgs/trilogy-group/teams/aurea-central-reviewers/members).

<details><summary>More information</summary>

_This is an initiative to improve overall code quality. If your change is **urgent**, you **don't** have to wait for our review._ 

_NOTE: This review **does not** replace your team's review._

</details>

File: DNN Platform/Library/ComponentModel/SimpleContainer.cs [hunk 1]
```diff
@@ -215,7 +215,7 @@ public override string[] GetComponentList(Type contractType)
         {
             var components = new List<string>();
 
-            using (_registeredComponents.GetReadLock())
+            using (_registeredComponents.GetUpgradeableReadLock())
```
Comment (leonardopires) [inline DNN Platform/Library/ComponentModel/SimpleContainer.cs]: :white_check_mark: It is nice to see that this code employs the `using` statement to ensure unmanaged resources get cleaned up properly.  
  
**No change suggested. Looks good.** 

<a id="23727700-93e8-11e8-ad1e-d3c2fa636bed" href="none" codereview-uuid-reserved-tag />

File: DNN Platform/Library/Collections/ReaderWriterLockStrategy.cs [hunk 2]
```diff
@@ -86,6 +86,24 @@ public ISharedCollectionLock GetReadLock(TimeSpan timeout)
...
+        public ISharedCollectionLock GetUpgradeableReadLock(TimeSpan timeout)
+        {
+            EnsureNotDisposed();
+            if (Lock.TryEnterUpgradeableReadLock(timeout))
+            {
+                return new ReaderWriterSlimLock(Lock);
+            }
+            else
+            {
+                throw new ApplicationException("ReaderWriterLockStrategy.GetUpgradeableReadLock timed out");
```
Comment (leonardopires) [inline DNN Platform/Library/Collections/ReaderWriterLockStrategy.cs]: :no_entry_sign: **This code throws an instance of a reserved exception.**  
  
An important improvement would be to use a specific exception type instead.  
  
This change would avoid any usage of [a reserved exception](https://msdn.microsoft.com/en-us/library/ms182338.aspx). Such kind of exception should only be thrown by the .NET Framework itself.  

**[More details](https://confluence.devfactory.com/x/4ROkEw)**
          
## The solution in this case

```cs  
throw new TimeoutException("ReaderWriterLockStrategy.GetUpgradeableReadLock timed out");
```  
> NOTE:  
> - Consider applying the structure of this solution in other parts of the code, if there is another occurrence. 

<a id="6a741b90-93e8-11e8-ad1e-d3c2fa636bed" href="none" codereview-uuid-reserved-tag />

File: DNN Platform/Library/Collections/ExclusiveLockStrategy.cs [hunk 3]
```diff
@@ -141,5 +141,15 @@ private void EnsureNotDisposed()
                 throw new ObjectDisposedException("ExclusiveLockStrategy");
             }
         }
+
+        public ISharedCollectionLock GetUpgradeableReadLock()
+        {
+            throw new NotImplementedException();
```
Comment (leonardopires) [inline DNN Platform/Library/Collections/ExclusiveLockStrategy.cs]: :no_entry_sign: **This type/member doesn't seem to have a meaningful implementation and/or the implementation is missing.**  
  
A possible improvement would be to remove this type/member or to write why this type/member was left without implementation.  
  
This change would ensure all types/members have a logical and meaningful implementation, and also reduce the complexity by making the code cleaner, or ensure that other developers would easily understand the reason that this type/member has no implementation.  

**[More details](https://confluence.devfactory.com/x/gAG8Ew)**
          
## The solution in this case

```cs  
/// <summary>
/// Decribe what the method was supposed to do.
/// </summary>
/// <remarks>
/// describe the decision here
/// why this type/member was left without implementation
/// </remarks>
```  
> NOTE:  
> - Consider applying the structure of this solution in other parts of the code, if there is another occurrence. 

<a id="924fa8f0-93e8-11e8-ad1e-d3c2fa636bed" href="none" codereview-uuid-reserved-tag />

**Comment by daguiler** [comment]
Closing. Accepted solution for this issue is https://github.com/dnnsoftware/Dnn.Platform/issues/2185.

**Pull request closed without merging**